lib/Dialect/TritonGPU/Transforms/AccelerateMatmul.cpp

```diff
@@ -380,6 +383,148 @@ static void decomposeMixedModeDotOp(ModuleOp mod, int computeCapability) {
+                                             /*versionMinor=*/0, warpsPerCTA,
+                                             CTALayout, instrShape);
+    auto newRetType = RankedTensorType::get(
+        oldRetType.getShape(), oldRetType.getElementType(), mmaEnc);
+
+    // convert accumulator
+    auto oldAcc = dotOp.getOperand(2);
+    auto newAcc =
+        rewriter.create<ConvertLayoutOp>(oldAcc.getLoc(), newRetType, oldAcc);
+
+    auto toMMABf16 = [&newRetType, &rewriter, &ctx,
+                      &enumToType](TypedValue<RankedTensorType> v, int idx,
+                                   F8F6F4Type type) {
+      // MMAv2 Layout
+      auto vType = v.getType();
+      auto newVEncoding = DotOperandEncodingAttr::get(
+          ctx, idx, newRetType.getEncoding(), enumToType((type)));
+      auto newVType = RankedTensorType::get(
+          v.getType().getShape(), v.getType().getElementType(), newVEncoding);
+      v = rewriter.create<ConvertLayoutOp>(v.getLoc(), newVType, v);
+
+      // Bitcast
+      auto vTypeFp8 = RankedTensorType::get(
+          vType.getShape(), rewriter.getFloat8E4M3FNType(), newVEncoding);
+      v = cast<TypedValue<RankedTensorType>>(
+          rewriter.create<BitcastOp>(v.getLoc(), vTypeFp8, v).getResult());
+
+      // Convert to bf16
+      auto vTypeBf16 = RankedTensorType::get(
+          vType.getShape(), rewriter.getBF16Type(), newVEncoding);
+      return rewriter.create<FpToFpOp>(v.getLoc(), vTypeBf16, v);
+    };
+    a = toMMABf16(a, 0, aType);
+    b = toMMABf16(b, 1, bType);
+
+    // [Note: A trick to avoid warp shuffles in the lowering]
+    // FIXME: Implement this when we can set general layouts on a tensor
+
+    // For bf16, we have 4 threads per row
+    // https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#mma-16816-a-f16
+    // and each of them needs to get every scale in that row.
+    // It turns out that the layout for the output of type bf16 gives us exactly
+    // this layout when the number of mxfp vectors is equal to two (K = 64)
+    // https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#mma-16816-c
+    // This can be generalised to other K with linear layouts, but the general
+    // layout cannot cannot be represented with the predefined layouts :(
+    // With this trick, we could do the full lowering here and remove the
+    // UpcastMXFPOp altogether
+
+    auto order =
+        cast<DotOperandEncodingAttr>(a.getType().getEncoding()).getCTAOrder();
+
+    // TODO Adjust for mmav3
+    // Necessary choice to leave all the scales of the tile in that given warp
+
+    auto shapeTileA = std::array<unsigned, 2>{instrShape[0], instrShape[0]};
+    auto threadsPerWarp =
+        SmallVector<unsigned>{shapeTileA[order[0]], 32 / shapeTileA[order[0]]};
+
+    // The number of scales in a given warp * 32 gives us the minimum K
+    if (a.getType().getShape()[order[0]] < threadsPerWarp[order[0]] * 32) {
+      return failure();
+    }
```

Comment:
I thought the minimum K would always be 32, why do we have this extra restriction?

Reply:
I think the exact formula I wrote here is wrong, but the idea is (assume `order[0] = 1` for the examples):
1. To have all the threads in a warp load scales that correspond to the values in a given tile, it needs to be of size `[tileShape[0]], threadsPerWarp[0] / tileShape[0]`. For example, for a tile of size `[16, 16]`, we need to choose for scale a `threadsPerWarp = [16, 2]`.
2. Running with the example above, if `threadsPerWarp = [16, 2]`, this means that each warp is loading two scales.  
3. These two scales make us having to load at least 2 mxfp vectors on the K dim, which makes K >= 64.

Reply:
I see, but in this example if K=32 we could still use [16, 2] and the scale would be duplicated and things would still work? 
Is this just restricted for simplicity as a first step? Maybe we should have a TODO?

Reply:
yes makes sense to start with the simple case